utils/node-loader/src/batch_pool.rs

```diff
@@ -248,24 +257,55 @@ async fn process_events(
     let mut mailbox_added = BTreeSet::new();
 
     let results = loop {
-        let r = match api.events_since(block_hash, wait_for_events_blocks).await {
-            Ok(mut v) => {
-                // `gclient::EventProcessor` implementation on `IntoIterator` implementor clones `self` without mutating
-                // it, although `proc` and `proc_many` take mutable reference on `self`.
-                let mut mailbox_from_events = utils::capture_mailbox_messages(&api, &mut v)
-                    .await
-                    .expect("always valid by definition");
-                mailbox_added.append(&mut mailbox_from_events);
-
-                v.err_or_succeed_batch(messages.keys().copied()).await
+        let mut events = rx.recv().await?;
+        while events.block_hash() != block_hash {
```

Comment:
what's the purpose of this while loop? why can't we rely only on values from `r` below?

Reply:
we need to get events since specific block hash. We are waiting for it here.